### PR TITLE
DATACOUCH-484 - Thread safety issue using findBy.

### PR DESCRIPTION
**Pull request by davidkelly**
Support for positional parameters necessitated caching them, however
that was not done in a threadsafe fashion.  ThreadLocal storage is
sufficient to take care of the issue.  Added test for it while at it.

### CODE REVIEW

File: src/main/java/org/springframework/data/couchbase/repository/query/PartTreeN1qlBasedQuery.java [hunk 1]
```diff
@@ -50,16 +50,21 @@
...
   private final PartTree partTree;
-  private JsonValue placeHolderValues;
+  private ThreadLocal<JsonValue> placeHolderValues;
 
   public PartTreeN1qlBasedQuery(CouchbaseQueryMethod queryMethod, CouchbaseOperations couchbaseOperations) {
     super(queryMethod, couchbaseOperations);
     this.partTree = new PartTree(queryMethod.getName(), queryMethod.getEntityInformation().getJavaType());
+    this.placeHolderValues = new ThreadLocal<JsonValue>() {
+      @Override public JsonValue initialValue() {
+        return JsonArray.create();
```
Comment (daschl) [inline src/main/java/org/springframework/data/couchbase/repository/query/PartTreeN1qlBasedQuery.java]: looks like this was previously uninitialized (the inner value) should this still be null or does this any harm? My assumption is no but still checking

File: src/test/java/org/springframework/data/couchbase/repository/N1qlCouchbaseRepositoryIntegrationTests.java [hunk 2]
```diff
@@ -90,6 +97,41 @@ public void cleanUp() {
...
+  public void doShouldBeThreadsafe() {
+    int threads = 50;
+    ExecutorService service = Executors.newFixedThreadPool(threads);
+    List<Callable<Boolean>> callables = new ArrayList<>();
+    for (int thread = 0; thread < threads; ++thread) {
+      final int counter = thread;
+      Callable<Boolean> booleanSupplier = () -> {
+        String expectedName = "party like it's 199" + counter%12;
+        String foundName = partyRepository.findByName(expectedName).get(0).getName();
+        System.out.println("expected = " + expectedName  + " found = " + foundName);
```
Comment (daschl) [inline src/test/java/org/springframework/data/couchbase/repository/N1qlCouchbaseRepositoryIntegrationTests.java]: make a real assertEquals and not println so it fails the test if needed and not spam the logs

Comment (davidkelly) [reply src/test/java/org/springframework/data/couchbase/repository/N1qlCouchbaseRepositoryIntegrationTests.java]: I can just remove this line -- I'd popped it in to make understanding what happened easier, during development.

**Comment by davidkelly** [comment]
@mp911de - The design that was there cached the parameters rather than keeping them on the stack as it passed through the various calls.  I felt like just making this cached value thread-local was safer than re-working the entire set of calls (that span quite a few objects).  

Part of my goal was getting a fix to a customer, so this seemed safer.  But certainly it isn't the optimal design.  Had I more time I'd rework the N1QL utility classes.  I suppose I can still do so, instead of this, but not right away.

**Comment by mp911de** [comment]
Thanks a lot for the details. Let's please address that issue for the 4.0 line.

**Comment by davidkelly** [comment]
@mp911de - how about we put this in master and get it out to the customer.  Then, in the new SDK3 version which is coming soon, we will do this right and keep the parameters on the call stack.  LMK what you think.

File: src/main/java/org/springframework/data/couchbase/repository/query/PartTreeN1qlBasedQuery.java (hunk 1 above)
Comment (davidkelly) [reply src/main/java/org/springframework/data/couchbase/repository/query/PartTreeN1qlBasedQuery.java]: This does no harm - IIRC not initializing it was a problem.

**Comment by mp911de** [comment]
Works for me @davidkelly, that's what I meant with let's fix it for the 4.0 (SDK3) line.

**Comment by davidkelly** [comment]
@mp911de Thx!  I think my browser hadn't noticed your comment before I made the second one 👍

**Comment by mikereiche** [comment]
This issue is still in 3.0.12-RELEASE through 3.2.2-RELEASE

**Comment by mp911de** [comment]
Please note that only the 3.1 and 3.2 lines (Lovelace and Moore) are maintained and actively released.

**Comment by mikereiche** [comment]
Thanks.  I checked and found the 413 issue in 3.1.0, 3.1.1 and 3.1.2 only.  I'll look at what change was made for 3.1.3.

**Pull request merged**